### ISSUE TITLE
Appropriate for framework?  Utility class for working with repeating instances

### ISSUE THREAD

**Issue by 123andy** [vanderbilt/redcap-external-modules#286]
So, working with repeating instances in REDCap is challenge... They return fields that are part of the repeating form, you have to call a method on the new Project object to get it to initialize... Getting things like the 'next instance' is a bit of a challenge...

So, we started a helper class to work with repeat instances and I'm wondering, where, if anywhere, such a helper class belongs.

1) We can repeat it and namespace it to each EM that uses it.  but, this gets a bit repetitve.
2) We could try to make it part of Framework for EMs but that is maybe a bit overkill since it isn't itself a static method
3) We could invent a way to add classes to redcap, like a 'type' of external module... Not sure where that goes...

Just thought I'd bring it up for discussion...

**Comment by grezniczek** [comment]
I have been asking myself the same question. Where best to put reusuable bits of code? Especially those that might benefit others in the community, too, so not everybody needs to reinvent the same wheels.

An idea was that such utility classes, functions, ... could be consolidated in the [External Module Templater](https://github.com/vanderbilt-redcap/external-module-templater) (EMT) GitHub repo (the deluxe-version would have integration with the GUI builder ... "Include XY library"). This would both, facilitate code-sharing (one doesn't necessarily have to use the EMT module) and prevent  version conflicts, as the utilities would always be scoped to the EM that uses them.

Alternatively, a separate (curated) GitHub repo could be established. Can we build a `redcap-compose`? Or just use composer and publish such classes with packagist. I've never done it, but it seems to be little effort? A naming convention could be established and a directory (just a README here on GitHub), that is linked to from the EM docs.

**Comment by 123andy** [comment]
Are you suggesting that when you make your EM template you could choose to include helper-library-x and it would pull that class into your template and namespace it?    So, your final EM would be a self-contained entity but you would get the benefit of having a library of helper classes? 

I like this idea!  Reminds me I've also make some modifications on the templater I need to push up...

**Comment by grezniczek** [comment]
Yes, that would be the idea if the EMT is chosen as the host for stuff like this.

**Comment by mmcev106** [comment]
This is something I've struggled with greatly... Even within my own team it's been very difficult to create a shared library that is more appealing that copy/pasting and reinventing the wheel.  I'm not entirely sure why...  The best success we've had so far is with the EM framework, but I still kind of feel like I'm on my own with it most of the time.  I also haven't had any incentive to make new module features available outside the context of a module since I always create new projects as modules rather than plugins...

It seems like PRs to add methods to the REDCap should have caught on long before external modules existed.  I wonder if we should just add more utility classes to REDCap core, and officially document and support them.

**Comment by mmcev106** [comment]
I guess it wouldn't make much difference whether new utility classes existed in REDCap core or this repo...  The latter might be easier logistically for PRs, unit testing, etc.

**Comment by grezniczek** [comment]
While working on the [Stopwatch EM](https://github.com/davwillev/stopwatch) and following @123andy's [suggestion](https://community.projectredcap.org/comments/83284/view.html) of storing time lapses in instances of a repeating form, there was, of course, the HUGE challenge of working with instance data. 

To make this a bit more manageable, I abstracted away some of the pain in two classes built around the _Project Data Structure_ already submitted as part of PR #307. 

Maybe you two, @mmcev106 and @123andy, could have a look at the `Project` and `Record` classes ([found here](https://github.com/davwillev/stopwatch/tree/master/classes)). The `Project` class is basically the one from the EF Framework, but incorporating the _Project Data Structure_ and several helper methods. The `Record` class concerns itself with handling loading and saving record data. This is what I needed to get the Stopwatch EM working and thus far from complete. But it might be a good start to include in the EM Framework (a PR would pretty much be a simple copy/paste effort). I guess the `Record` class could be integrated with the `Records` class of the EM Framework (which as far as I can tell only handles getting lock status and locking of records).

**Comment by mmcev106** [comment]
Those classes look great.  I'd happy merge a PR with them.

**Comment by grezniczek** [comment]
I'll probaby get around to submitting a PR after the weekend. This will then supersede #307.
